### PR TITLE
meta: replace predicates with pattern over elaborated platform

### LINKED ISSUE #212925: `isStatic` has no non-function representation, but `nix-env --meta` assumes one exists
I am working on this bug.  First reported [here](https://github.com/NixOS/nixpkgs/pull/192197#issuecomment-1406185236).

The root cause of the problem is that there is no "pattern" (see `lib/system/inspect.nix`) representation for `isStatic`.  It is special, unlike all the other platform predicates.  It is this way because `pkgsStatic` exists, and because

```
stdenv.hostPlatform.config == pkgsStatic.stdenv.hostPlatform.config
```

... in other words, "static-ness" is not represented in the `gnu-config` triple for a platform.

Right now we try to pretend that you can reconstruct all the pattern-representable aspects of a `hostPlatform` from its `config`.  That isn't really true, because of `isStatic` and `useLLVM`, and it's getting more untrue as time goes on.  The long-term solution is to stop pretending.

I have a possible short-term fix that I'm testing.

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

Putting functions in `meta.badPlatforms` is, well, bad.

I [worried about this](https://github.com/NixOS/nixpkgs/pull/192197#issuecomment-1403035046) a [bit](https://github.com/NixOS/nixpkgs/pull/192197#issuecomment-1403571073), and then it [became problematic](https://github.com/NixOS/nixpkgs/pull/192197#issuecomment-1406185236) sooner than I expected it would.

This removes the ability to put functions in `meta.badPlatforms` (closes #212925).

In order to preserve the ability to use `isStatic` there, it expands `meta.platformMatch` to allow matching over the entire (elaborated) `hostPlatform` rather than just  `hostPlatform.parsed`.

cc:
- @ncfavier for that review
- @Ericson2314 for a chance to object on philosophical grounds
- @alyssais who uses `isStatic`

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by ncfavier — COMMENTED** [review]
Seems reasonable enough

File: lib/systems/inspect.nix [hunk 1]
```diff
@@ -90,4 +91,13 @@ rec {
...
 
   predicates = mapAttrs (_: matchAnyAttrs) patterns;
+
+  # these patterns are to be matched against the entire
+  # {host,build,target}Platform structure; they include a `parsed={}` marker so
+  # that `lib.meta.availableOn` can distinguish them from the patterns which
+  # apply only to the `parsed` field.
+
+  platformPatterns = {
+    isStatic = { parsed = {}; isStatic = true; };
```
Comment (ncfavier) [inline lib/systems/inspect.nix:101]:
```suggestion
  platformPatterns = mapAttrs (p: { parsed = {}; } // p) {
    isStatic = { isStatic = true; };
```

Maybe?

Comment (amjoseph-nixpkgs) [reply lib/systems/inspect.nix:101]: Definitely.

**Comment by alyssais** [comment]
I'm basically resigned to not being able to put `isStatic` into `platforms`/`badPlatforms` at this point, so this is fine with me.  Would be cool if we could figure out how to express it as a pattern though.

**Pull request merged**